File: Dokie/ViewControllers/Document/ImagesPageView/DocumentImagesPageViewController.m
//
//  DocumentImagesPageViewController.m
//  Dokie
//
//  Created by Ashamaz Shidov on 26.01.2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

#import "DocumentImagesPageViewController.h"
#import "DocumentImagesPageViewModel.h"
#import "Logger.h"
#import "ImagePreviewViewController.h"
#import "FileImageView.h"
#import "AppearanceManager.h"

#import <ReactiveObjC/ReactiveObjC.h>

@interface DocumentImagesPageViewController() <UIGestureRecognizerDelegate, UIPageViewControllerDelegate,UIPageViewControllerDataSource>

@property (nonatomic, strong) DocumentImagesPageViewModel *viewModel;

@property (nonatomic, strong) UIPanGestureRecognizer *panGesture;
@property (nonatomic) CGPoint startLocation;
@property (nonatomic) CGFloat deltaY;
@property (nonatomic) CGFloat lastDirection;

@property (nonatomic, strong) ImagePreviewViewController *currentVisibleViewController;

@end

@implementation DocumentImagesPageViewController

- (instancetype)initWithViewModel:(DocumentImagesPageViewModel *)viewModel {
    ParameterAssert(viewModel);
    
    self = [self initWithTransitionStyle:UIPageViewControllerTransitionStyleScroll navigationOrientation:UIPageViewControllerNavigationOrientationHorizontal options:nil];

    if (self) {
        self.viewModel = viewModel;
        self.dataSource = self;
        self.delegate = self;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setup];
}

- (void)setup {
    self.view.backgroundColor = [AppearanceManager backgroundColor];
    @weakify(self);
    UIBarButtonItem *closeButton = [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"Close", @"Close navigation bar button text") style:UIBarButtonItemStylePlain target:nil action:nil];
    self.navigationItem.leftBarButtonItems = @[closeButton];
    closeButton.rac_command = [[RACCommand alloc] initWithSignalBlock:^RACSignal * _Nonnull(id _) {
        @strongify(self);
        [self closePressed];
        return [RACSignal empty];
    }];
    
    UIBarButtonItem *shareButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAction target:nil action:nil];
    self.navigationItem.rightBarButtonItems = @[shareButton];
    shareButton.rac_command = [[RACCommand alloc] initWithSignalBlock:^RACSignal * _Nonnull(id _) {
        @strongify(self);
        if (self.currentVisibleViewController.imageView.image) {
            UIActivityViewController *avc = [[UIActivityViewController alloc] initWithActivityItems:@[self.currentVisibleViewController.imageView.image] applicationActivities:nil];
            [self presentViewController:avc animated:YES completion:nil];
        }
        return [RACSignal empty];
    }];
    
    [[RACObserve(self, viewModel.contentImages) distinctUntilChanged] subscribeNext:^(id _) {
        @strongify(self);
        if (self.currentVisibleViewController.pageIndex >= self.viewModel.contentImages.count || !self.currentVisibleViewController) {
            self.currentVisibleViewController = [self firstViewController];
        }
        else {
            self.currentVisibleViewController = [self imagePreviewViewControllerForPageIndex:self.currentVisibleViewController.pageIndex];
        }
        if (self.currentVisibleViewController) {
            self.navigationItem.title = [NSString stringWithFormat:@"%ld / %ld", (long)self.currentVisibleViewController.pageIndex+1, (long)self.viewModel.contentImages.count];
            [self setViewControllers:@[self.currentVisibleViewController]
                           direction:UIPageViewControllerNavigationDirectionForward
                            animated:YES
                          completion:NULL];
        }
    }];
    
    self.panGesture = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(panGesture:)];
    self.panGesture.delegate = self;
    self.panGesture.delaysTouchesBegan = YES;
    [self.view addGestureRecognizer:self.panGesture];
}

- (ImagePreviewViewController *)firstViewController {
    if (self.viewModel.contentImages.count == 0) {
        return nil;
    }
    else if (self.viewModel.firstIndex < self.viewModel.contentImages.count) {
        return [self imagePreviewViewControllerForPageIndex:self.viewModel.firstIndex];
    }
    else {
        return [self imagePreviewViewControllerForPageIndex:0];
    }
}

- (void)hideWindow {
    [self.navigationController dismissViewControllerAnimated:YES completion:nil];
}

- (void)closePressed {
    [UIView animateWithDuration:0.3
                          delay:0
                        options:UIViewAnimationOptionCurveLinear
                     animations:^{
                         CGPoint finalPoint = CGPointMake(self.navigationController.view.center.x, self.navigationController.view.frame.size.height*1.5 + 100);
                         self.view.center = finalPoint;
                         self.navigationController.view.alpha = 0.1;
                     }
                     completion:^(BOOL finished) {
                         if (finished) {
                             [self hideWindow];
                         }
                     }];
}

- (void)moveScrollViewCenterToDeltaY:(CGFloat)deltaY {
    if (self.currentVisibleViewController)
        [self.currentVisibleViewController moveScrollViewCenterToDeltaY:deltaY];
}

- (void)panGesture:(UIPanGestureRecognizer *)sender {
    if (sender.state == UIGestureRecognizerStateBegan) {
        self.startLocation = [sender translationInView:self.view];
        self.deltaY = 0;
        self.lastDirection = 0;
    }

    CGPoint translation = [sender translationInView:self.view];
    self.deltaY += translation.y;
    if (translation.y != 0 && fabs(translation.y) > 10) {
        self.lastDirection = translation.y;
    }
    [self moveScrollViewCenterToDeltaY:self.deltaY];

    self.navigationController.view.alpha = 1.1 - fabs(2*self.deltaY)/self.view.frame.size.height;
    
    if (sender.state == UIGestureRecognizerStateEnded) {
        BOOL movedUp = self.startLocation.y > self.deltaY;
        BOOL lastDirectionIsUp = self.lastDirection < 0;
        if (fabs(self.deltaY) > sender.view.frame.size.height/6 &&
            (movedUp == lastDirectionIsUp)) {
            [UIView animateWithDuration:0.3
                                  delay:0
                                options:UIViewAnimationOptionCurveLinear
                             animations:^{
                                 CGPoint finalPoint;
                                 if (movedUp) {
                                     finalPoint = CGPointMake(self.navigationController.view.center.x, -(sender.view.frame.size.height + 100));
                                 }
                                 else {
                                     finalPoint = CGPointMake(self.navigationController.view.center.x,   sender.view.frame.size.height + 100);
                                 }
                                 [self moveScrollViewCenterToDeltaY:finalPoint.y];
                                 self.navigationController.view.alpha = 0.1;
                             }
                             completion:^(BOOL finished) {
                                 if (finished) {
                                     [self hideWindow];
                                 }
                             }];

        }
        else {
            [UIView animateWithDuration:0.3
                                  delay:0
                                options:UIViewAnimationOptionCurveLinear
                             animations:^{
                                 [self moveScrollViewCenterToDeltaY:0];
                                 self.navigationController.view.alpha = 1;
                             }
                             completion:nil];
        }
    }
    
    [sender setTranslation:CGPointMake(0, 0) inView:self.view];
}

#pragma mark - UIPageViewControllerDelegate

- (void)pageViewController:(UIPageViewController *)pageViewController didFinishAnimating:(BOOL)finished previousViewControllers:(NSArray<UIViewController *> *)previousViewControllers transitionCompleted:(BOOL)completed {
    if (completed) {
        self.currentVisibleViewController = pageViewController.viewControllers.firstObject;
        self.navigationItem.title = [NSString stringWithFormat:@"%ld / %ld", (long)(self.currentVisibleViewController.pageIndex+1), (unsigned long)self.viewModel.contentImages.count];
    }
}

#pragma mark - UIPageViewControllerDataSource

- (ImagePreviewViewController *)imagePreviewViewControllerForPageIndex:(NSInteger)index {
    if (self.viewModel.contentImages.count <= index || !self.viewModel.contentImages) {
        return nil;
    }
    
    CDContent *content = self.viewModel.contentImages[index];
    FileImageView *imageView = [FileImageView new];
    imageView.translatesAutoresizingMaskIntoConstraints = NO;
    imageView.contentMode = UIViewContentModeScaleAspectFit;
    [imageView updateWithContent:content];
    ImagePreviewViewController *viewController = [[ImagePreviewViewController alloc] initWithImageView:imageView pageIndex:index];
    return viewController;
}

- (UIViewController *)pageViewController:(UIPageViewController *)pvc viewControllerBeforeViewController:(ImagePreviewViewController *)vc {
    NSUInteger index = vc.pageIndex;
    return [self imagePreviewViewControllerForPageIndex:(index - 1)];
}

- (UIViewController *)pageViewController:(UIPageViewController *)pvc viewControllerAfterViewController:(ImagePreviewViewController *)vc {
    NSUInteger index = vc.pageIndex;
    return [self imagePreviewViewControllerForPageIndex:(index + 1)];
}

#pragma mark - UIGestureRecognizerDelegate

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer {
    if ([otherGestureRecognizer isKindOfClass:[UIPinchGestureRecognizer class]] || [otherGestureRecognizer isKindOfClass:[UIPanGestureRecognizer class]])
        return NO;
    return YES;
}

@end

File: Dokie/ViewControllers/Document/ImagesPageView/DocumentImagesPageViewController.swift
//
//  DocumentImagesPageViewController.swift
//  Dokie
//
//  Created by Ashamaz Shidov on 25.05.2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

import UIKit
import ReactiveCocoa
import ReactiveSwift

class DocumentImagesPageViewController: UIPageViewController {
    
    private let viewModel: DocumentImagesPageViewModel!
    private var startLocation = CGPoint()
    private var deltaY = CGFloat(0)
    private var lastDirection = CGFloat(0)
    
    private var currentVisibleViewController: ImagePreviewViewController?
    
    init(viewModel: DocumentImagesPageViewModel) {
        self.viewModel = viewModel
        super.init(transitionStyle: .scroll, navigationOrientation: .horizontal, options: nil)
        delegate = self
        dataSource = self
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setup()
    }
    
    private func setup() {
        view.backgroundColor = Appearance.backgroundColor
        
        let closeButton = UIBarButtonItem(title: String.localized("Close"), style: .plain, target: self, action: #selector(closePressed))
        navigationItem.leftBarButtonItems = [closeButton]
        
        let shareButton = UIBarButtonItem(barButtonSystemItem: .action, target: self, action: #selector(share))
        navigationItem.rightBarButtonItems = [shareButton]
        
        viewModel
            .contentImages
            .producer
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithValues { [unowned self] images in
                if let currentVisibleViewController = self.currentVisibleViewController,
                    currentVisibleViewController.pageIndex < self.viewModel.contentImages.value.count {
                    self.currentVisibleViewController = self.imagePreviewViewController(forPage: currentVisibleViewController.pageIndex)
                }
                else {
                    self.currentVisibleViewController = self.firstViewController()
                }
                
                guard let currentVisibleViewController = self.currentVisibleViewController else { return }
                
                self.navigationItem.title = "\(currentVisibleViewController.pageIndex+1) / \(self.viewModel.contentImages.value.count)"
                self.setViewControllers([currentVisibleViewController], direction: .forward, animated: true, completion: nil)
        }
        
        let panGesture = UIPanGestureRecognizer(target: self, action: #selector(panGesture(sender:)))
        panGesture.delegate = self
        panGesture.delaysTouchesBegan = true
        view.addGestureRecognizer(panGesture)
    }
    
    private func firstViewController() -> ImagePreviewViewController? {
        if viewModel.contentImages.value.count == 0 {
            return nil
        }
        else if viewModel.firstIndex < viewModel.contentImages.value.count {
            return imagePreviewViewController(forPage: viewModel.firstIndex)
        }
        else {
            return imagePreviewViewController(forPage: 0)
        }
    }
    
    private func imagePreviewViewController(forPage index: Int) -> ImagePreviewViewController? {
        guard index < viewModel.contentImages.value.count,
            index >= 0 else {
            return nil
        }
        
        let content = viewModel.contentImages.value[index]
        let imageView = FileImageView()
        imageView.translatesAutoresizingMaskIntoConstraints = false
        imageView.contentMode = .scaleAspectFit
        imageView.update(content: content)
        return ImagePreviewViewController(imageView: imageView, pageIndex: index)
    }
    
    @objc func share() {
        if let image = currentVisibleViewController?.imageView.image {
            let avc = UIActivityViewController(activityItems: [image], applicationActivities: nil)
            present(avc, animated: true, completion: nil)
        }
    }
    
    @objc func closePressed() {
        UIView.animate(withDuration: 0.3,
                       delay: 0,
                       options: .curveLinear,
                       animations: { [weak self] in
                        let finalPoint = CGPoint(x: self?.navigationController?.view.center.x ?? 0, y: self?.navigationController?.view.frame.size.height ?? 0*1.5 + 100)
                        self?.view.center = finalPoint
                        self?.navigationController?.view.alpha = 0.1
        }) { [weak self] finished in
            if (finished) {
                self?.hideWindow()
            }
        }
    }
    
    @objc func panGesture(sender: UIPanGestureRecognizer) {
        defer {
            sender.setTranslation(CGPoint(x: 0, y: 0), in: view)
        }
        
        if sender.state == .began {
            startLocation = sender.location(in: view)
            deltaY = 0
            lastDirection = 0
        }
        
        let translation = sender.translation(in: view)
        deltaY += translation.y
        if translation.y != 0 && abs(translation.y) > 10 {
            lastDirection = translation.y
        }
        moveScrollViewCenter(delta: deltaY)
        
        navigationController?.view.alpha = 1.1 - abs(2*deltaY)/view.frame.height
        
        guard let senderViewFrame = sender.view?.frame else { return }
        
        if sender.state == .ended {
            let movedUp = startLocation.y > deltaY
            let lastDirectionIsUp = lastDirection < 0
            if abs(deltaY) > senderViewFrame.height/6,
                movedUp == lastDirectionIsUp {
                UIView.animate(withDuration: 0.3, delay: 0, options: .curveEaseIn, animations: { [unowned self] in
                    let finalPoint: CGPoint
                    if movedUp {
                        finalPoint = CGPoint(x: self.navigationController?.view.center.x ?? 0, y: -(senderViewFrame.height + 100))
                    }
                    else {
                        finalPoint = CGPoint(x: self.navigationController?.view.center.x ?? 0, y: senderViewFrame.height + 100)
                    }
                    self.moveScrollViewCenter(delta: finalPoint.y)
                    self.navigationController?.view.alpha = 0
                }) { [unowned self] finished in
                    if finished {
                        self.hideWindow()
                    }
                }
            }
            else {
                UIView.animate(withDuration: 0.3, delay: 0, options: .curveEaseIn, animations: { [unowned self] in
                    self.moveScrollViewCenter(delta: 0)
                    self.navigationController?.view.alpha = 1
                    }, completion: nil)
            }
        }
    }
    
    private func hideWindow() {
        navigationController?.dismiss(animated: true, completion: nil)
    }
    
    private func moveScrollViewCenter(delta: CGFloat) {
        currentVisibleViewController?.moveScrollViewCenter(delta: delta)
    }
    
}

extension DocumentImagesPageViewController : UIGestureRecognizerDelegate {
    
    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
        if otherGestureRecognizer.isKind(of: UIPinchGestureRecognizer.self) || otherGestureRecognizer.isKind(of: UIPanGestureRecognizer.self) {
            return false
        }
        return true
    }
    
}

extension DocumentImagesPageViewController : UIPageViewControllerDelegate, UIPageViewControllerDataSource {
    
    func pageViewController(_ pageViewController: UIPageViewController, didFinishAnimating finished: Bool, previousViewControllers: [UIViewController], transitionCompleted completed: Bool) {
        currentVisibleViewController = pageViewController.viewControllers?.first as? ImagePreviewViewController
        if let currentVisibleViewController = currentVisibleViewController {
            navigationItem.title = "\(currentVisibleViewController.pageIndex + 1) / \(viewModel.contentImages.value.count)"
        }
    }
    
    func pageViewController(_ pageViewController: UIPageViewController, viewControllerBefore viewController: UIViewController) -> UIViewController? {
        if let beforeVC = viewController as? ImagePreviewViewController {
            return imagePreviewViewController(forPage: beforeVC.pageIndex - 1)
        }
        
        return nil
    }
    
    func pageViewController(_ pageViewController: UIPageViewController, viewControllerAfter viewController: UIViewController) -> UIViewController? {
        if let beforeVC = viewController as? ImagePreviewViewController {
            return imagePreviewViewController(forPage: beforeVC.pageIndex + 1)
        }
        
        return nil
    }
    
}
